students/johnn/session09/mailroom.py

```diff
@@ -0,0 +1,688 @@
+        In [596]:    
+
+        In [636]: d=Donor(full_name="john adams")
+        In [637]: d.add_donation=1000
+        In [638]: repr(d)
+        Out[638]: "Donor( id='7e93d724-de25-11e7-bee5-0800274b0a84', first_name='John', 
+            last_name='Adams', donations=[{'amount': 1000.0, 'date': '2017-12-11Z'}, 
+            {'amount': 1000.0, 'date': '2017-12-11Z'}], created='2017-12-11T03:43:47.686457Z' )"
+        In [639]:
+
+    """
+
+    def __init__(self, 
+            id="", created="",
+            full_name="",
+            first_name="", last_name="", middle_name="", suffix="",
+            donations=None, donation=None):
+
+        # normally no id is passed in, so we create one
+        if id == "":
+            id = str(uuid.uuid1())
+        # create a uuid for each record
+        self._id = id
+
+        # normally, no timestamp is passed in, so we create one
+        if created == "":
+            created = datetime.datetime.utcnow().isoformat() + "Z"
+        # keep track of record creation
+        self.created = created
+
+        # test to see if any of the name components are set
+        sub_name_set = any(sub_name != "" for sub_name in (
+            first_name,
+            middle_name,
+            last_name,
+            suffix))
+     
+        # we don't expect full_name and a subset to be set at the same time
+        if full_name != "" and sub_name_set:
+            raise ValueError("You cannot set 'full_name' and a subset of the name at the same time.")
+
+        # set the name via full_name or via the constituent parts
+        if full_name != "":
+            self.full_name = full_name
+        else:
+            self.first_name = first_name.title()
+            self.last_name = last_name.title()
+            self.middle_name = middle_name.title()
+            self.suffix = suffix.upper()
+
+        if (donations != None) and (donation != None):
+            raise ValueError("You cannot set 'donations' and 'donation' at the same time.")
+
+        if donation != None:
+            # if a donation is passed in, initialize the donations, then add the donation
+            self._donations = list()
+            self.add_donation(donation)
+        else:
+            if donations != None:
+                # TODO donations need more validation
+                self._donations = donations
+            else:
+                self._donations = list()
+            
+    @property
+    def id(self):
+        """ return the donor id """
+        return self._id
+
+    @id.setter
+    def id(self, value):
+        """ donor id cannot be changed once established """
+        raise AttributeError("You cannot change the 'id' after the record has been created.")
+
+    @property
+    def donations(self):
+        """ print all donations """
+        return self._donations
+
+    @donations.setter
+    def donations(self, value):
+        """ allow bulk setting of donation entries """
+        self._donations = value
```

Comment:
you probably want to wrap a list() around it -- so you'll get a copy, and be assured that it's a list, rather than, say, a tuple:

```
self._donations = list(value)
```
and better to use a meaningful name, like, day "donations"